src/python/pants/backend/native/tasks/link_object_files.py

```diff
@@ -194,9 +213,79 @@ def _execute_link_request(self, link_request):
       rc = process.wait()
       if rc != 0:
         workunit.set_outcome(WorkUnit.FAILURE)
-        raise self.LinkSharedLibrariesError(
+        raise self.LinkObjectFilesError(
           "Error linking native objects with command {cmd} and environment {env} "
           "for request {req}. Exit code was: {rc}."
           .format(cmd=cmd, env=env, req=link_request, rc=rc))
 
-    return SharedLibrary(name=native_artifact.lib_name, path=resulting_shared_lib_path)
+    return self.collect_output(native_artifact, output_file_path)
+
+
+class LinkSharedLibraries(LinkObjectFiles):
+  # TODO: validate that all shared libraries that a target depends on have a unique name! Otherwise
+  # we can't link them with `-l` in the same command line!
+
+  @classmethod
+  def product_types(cls):
+    return [SharedLibrary]
+
+  @classproperty
+  def linker_args(cls):
+    return {
+      'darwin': ['-Wl,-dylib'],
+      'linux': ['-shared'],
+    }
+
+  @memoized_method
+  def get_output_filename(self, native_artifact):
+    return native_artifact.as_shared_lib(self.platform)
+
+  def retrieve_output_from_cache(self, vt):
+    native_artifact = vt.target.ctypes_native_library
+    path_to_cached_lib = os.path.join(vt.results_dir, self.get_output_filename(native_artifact))
+    if not os.path.isfile(path_to_cached_lib):
+      raise self.LinkObjectFilesError("The shared library at {} does not exist!"
+                                          .format(path_to_cached_lib))
+    return SharedLibrary(name=native_artifact.lib_name, path=path_to_cached_lib)
+
+  def collect_output(self, native_artifact, output_file):
+    return SharedLibrary(name=native_artifact.lib_name, path=output_file)
+
+  def process_result(self, vt, result):
+    shared_libs_product = self.context.products.get(SharedLibrary)
+    self._add_product_at_target_base(shared_libs_product, vt.target, result)
+
+
+class LinkBinaries(LinkObjectFiles):
+
+  @classproperty
+  def linker_args(cls):
+    return {
+      'darwin': ['-Wl,-execute'],
+      'linux': [],
+    }
+
+  @memoized_method
+  def get_output_filename(self, native_artifact):
+    return native_artifact.exe_name
+
+  def native_artifact_targets(self):
+    return set(super(LinkBinaries, self).native_artifact_targets()) & set(self.context.target_roots)
```

Comment:
Is there no specific `cpp_binary` target type? Should there be?

With regard to including `self.context.target_roots` here: a pattern from `binary.jvm` is to privately produce (under `.pants.d`) binaries for all relevant targets in a context, but to only "publish" into `dist` the binaries for the target roots. This may not apply if there is not a specific `cpp_binary` target, but it feels a bit odd to me that there isn't.